### PR TITLE
Upgrade to TypeScript v4.4.2

### LINKED ISSUE #5355: Upgrade to TypeScript v4.4.2
Upgrade TypeScript to fix type errors in #5351 and to get access to new features such as better catch error types and improved control flow analysis (see [here](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#use-unknown-catch-variables)).

See [this conversation](https://parabol.slack.com/archives/C836NA350/p1630502503048500) in Slack for further context.

### PR DESCRIPTION
**Pull request by nickoferrall**
Fixes #5355 

Most changes are due to the catch error update: error is now `unknown` by default instead of `any`: https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#use-unknown-catch-variables

### CODE REVIEW

File: packages/client/utils/getDraftCoords.ts
Comment (nickoferrall) [inline packages/client/utils/getDraftCoords.ts:12]: It looks like `FakeClientRect` has the same properties as `ClientRect` except the `height` and `width` values are `readonly` in `ClientRect`. Where `getDraftCoords` is being used, the return value is often being used as type `ClientRect` so I think it's best to cast it here. For example, `cachedCoords` is being used [here](https://github.com/ParabolInc/parabol/blob/df50596b028bd866e4800b36cf36f0b034db5521/packages/client/components/TaskEditor/withEmojis.tsx#L95) as the `originCoords` prop which is of type `ClientRect`.

File: packages/client/components/VideoControls.tsx
Comment (nickoferrall) [inline packages/client/components/VideoControls.tsx:51]: This previously returned the error:
```
Conversion of type '("camera" | "microphone")[]' to type 'PermissionName[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
```
I think `Extract` is cleaner than casting it to `unknown` and then to `PermissionName`

**Comment by nickoferrall** [comment]
Señor @BartoszJarocki, I see that you liked the issue 😄  Would you mind reviewing this PR when you get the chance?

File: packages/client/components/VideoControls.tsx
Comment (BartoszJarocki) [reply packages/client/components/VideoControls.tsx:51]: This is funny. It's actually a bug in TS. 
https://github.com/microsoft/TypeScript-DOM-lib-generator/blob/main/baselines/dom.generated.d.ts#L17613 - these permissions are simply missing that's why casting fails. If they were present in the type definition the simplest solution would be to use `const descriptors = ['camera', 'microphone'] as const` - https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-inference 

TS has some issues with these types for some reason. That kind of issues happen from time to time: https://github.com/microsoft/TypeScript/issues/33923

File: packages/client/hooks/usePortal.tsx
Comment (BartoszJarocki) [inline packages/client/hooks/usePortal.tsx:36]: I'm not sure what's the right thing to do here because I'm pretty sure this code only runs in the browser (or am I wrong?). If that's the case then the TS is wrong here and the only way to let TS know we want the browser version of set timeout is to use `window.setTimeout` directly which is pretty dumb. Anyway, I'm sharing my thoughts here 😅

File: packages/server/graphql/handleGraphQLTrebuchetRequest.ts
Comment (BartoszJarocki) [inline packages/server/graphql/handleGraphQLTrebuchetRequest.ts:61]: To avoid nested if statements I'd change the first conditional to `e instanceof Error && e.message === 'TIMEOUT'`

File: packages/server/graphql/intranetSchema/mutations/draftEnterpriseInvoice.ts
Comment (BartoszJarocki) [inline packages/server/graphql/intranetSchema/mutations/draftEnterpriseInvoice.ts:126]: Not a big deal and probably depends on personal preferences but this could be simplified to 
```
      return {
        error: {message: e instanceof Error ? e.message : 'Unable to get billing leader user'}
      }
```
or 
```
      const message = e instanceof Error ? e.message : 'Unable to get billing leader user'
      return {
        error: {message}
      }
```

I don't have any strong preferences here so feel free to do whatever you think is best 👍🏻

File: packages/server/graphql/intranetSchema/mutations/loginSAML.ts
Comment (BartoszJarocki) [inline packages/server/graphql/intranetSchema/mutations/loginSAML.ts:64]: Same as above

**Review by BartoszJarocki — COMMENTED** [review]
Nice work! I've left a couple of comments to consider.

File: packages/client/hooks/usePortal.tsx
Comment (nickoferrall) [reply packages/client/hooks/usePortal.tsx]: `timeoutRef` was causing a typecheck issue but it looks like upgrading to v4.4.2 has fixed this so I've reverted `usePortal` back to the way it was

File: packages/client/components/VideoControls.tsx
Comment (nickoferrall) [reply packages/client/components/VideoControls.tsx]: Nice find! I've pushed an update adding a comment with a link to an issue and I'm now casting the name rather than using `Extract` (I think it's a tiny bit cleaner). I was tempted to create a PR to fix the issue but it looks like camera & microphone are considered [at risk](https://w3c.github.io/permissions/#dom-permissionname-camera)

**Comment by nickoferrall** [comment]
Thanks for the review @BartoszJarocki - I've pushed an update! Please let me know if you're happy with the changes and, if so, I'll move it along to maintainer review

**Review by BartoszJarocki — APPROVED** [review]
Nice work!

File: packages/server/graphql/mutations/pokerSetFinalScore.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/pokerSetFinalScore.ts:141]: missing return statement?
for whatever reason, `Error` instances can get swallowed & unreported (maybe their stack trace is too long? :shrug: ), so I always just return a POJO.
We should return `{error: {message}}` here

File: packages/server/graphql/mutations/setTaskEstimate.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/setTaskEstimate.ts:128]: let's return a POJO here instead of an Error instance

File: packages/server/graphql/mutations/upgradeToPro.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/upgradeToPro.ts:55]: let's put this in a conditional like
```ts
if ('param' in e) {
  const error = new Error(e.param)
  return ...
} else {
 ...
}
```

File: packages/server/utils/sendToSentry.ts
Comment (mattkrick) [inline packages/server/utils/sendToSentry.ts:14]: let's remove `unknown` from the input union
Some objects don't serialize & if that happens we'll be left with an empty error message

getting into the weeds, `JSON.stringify` simply calls an objects `toJSON` method (or the prototypical parent)
That means I can break it real easy (and many libs do, for good reasons)

e.g.
```ts
JSON.stringify({error: 'very important', toJSON: () => 'eat me'})
```

File: packages/server/utils/standardError.ts
Comment (mattkrick) [inline packages/server/utils/standardError.ts:3]: remove `unknown` union member

**Review by mattkrick — CHANGES_REQUESTED** [review]
thanks so much for tackling this so quickly.
couple minor things, good to merge after the little fixes!

File: packages/server/utils/sendToSentry.ts
Comment (nickoferrall) [reply packages/server/utils/sendToSentry.ts:14]: Ahh that's good to know! I was trying to retain the error and save code by avoiding:
```
catch(e){
    const message = e instanceof Error ? e.message : 'Custom error message'
    sendToSentry(new Error(message))
```
in each catch that has a sendToSentry call but it looks like that's the best bet

File: packages/server/graphql/mutations/upgradeToPro.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/upgradeToPro.ts:55]: Ah I tried to implement this but TypeScript complains that `Object is of type 'unknown'` on `if ('param' in e)`. If the two `any` casts in the catch are too ugly, an alternate option could be disabling an eslint rule that prevent us from doing `catch(e as any){`?

File: packages/server/graphql/mutations/pokerSetFinalScore.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/pokerSetFinalScore.ts]: Interesting, that's good to know! It looks like [standardError](https://github.com/ParabolInc/parabol/blob/3ea2c761bdaa507c0319b6f0da83435b43ecfe51/packages/server/utils/standardError.ts) returns a POJO (when the return statement isn't missing 😅 ) and sends the error to Sentry. Still, I've pushed a change returning `{error: {message}}` as we weren't previously sending the error to Sentry

**Comment by nickoferrall** [comment]
Thanks for the review @mattkrick! I've pushed an update which removes the unknown union type from sendToSentry. This resulted in quite a few changes so I'd prefer to hold off on merging until you've had the chance to scan through the updates

**Comment by mattkrick** [comment]
LGTM!

**Pull request merged**
